### PR TITLE
[PIP 79][common,broker,client] Change PartitionedTopicStats to aggregate by producerName

### PR DESCRIPTION
**Pull request by equanz**
Master Issue: https://github.com/apache/pulsar/wiki/PIP-79%3A-Reduce-redundant-producers-from-partitioned-producer

### Motivation

This is an alternative approach of https://github.com/apache/pulsar/pull/10534.
Also, please see the PIP document.

### Modifications

* Change partitioned producer initialization logic to use same producerName across partitions even if producerName is not configured at producer config
* Change PartitionedTopicStats to aggregate by producerName
  - To support partial partitioned producer like [PartialRoundRobinMessageRouterImpl.java](https://github.com/apache/pulsar/blob/1f8945addb18bf1f94366d2c7f5286ee6f62c675/pulsar-client/src/main/java/org/apache/pulsar/client/impl/customroute/PartialRoundRobinMessageRouterImpl.java)
  - Also, single producer which connected to partition are supported

Currently, this feature is enabled at all clients. It will change existing behavior (especially, total partitioned producer). If it is not approved, I'll add configurable interface to control partitioned producer initialization logic.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:
  - Added unit tests for stats aggregation

### Does this pull request potentially affect one of the following parts:
  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (yes)
    - Added new field `partial_producer_supported` to `CommandProducer` message
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation
- [x] `no-need-doc`

If we should add new stats aggregation logic, I'll modify to `doc-required`.

### CODE REVIEW

**Comment by nkurihar** [comment]
@merlimat @sijie @eolivelli 
Could you review if you have time?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApiTest2.java [hunk 1]
```diff
@@ -2231,4 +2232,170 @@ public void testFailedUpdatePartitionedTopic() throws Exception {
...
         assertNotNull(admin.topics().getStats(partitionedTopicName + "-partition-" + 6).getSubscriptions().get(subName1));
     }
+
+    @Test
+    public void testPartitionedStatsAggregationByProducerName() throws Exception {
+        final String topic = "persistent://prop-xyz/ns1/test-partitioned-stats-aggregation-by-producer-name";
+        admin.topics().createPartitionedTopic(topic, 10);
+
+        @Cleanup
+        PulsarClient client = PulsarClient.builder().serviceUrl(pulsar.getWebServiceAddress()).build();
```
Comment (massakam) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApiTest2.java:2242]: Is it not possible to use `pulsarClient` declared in the super class?
https://github.com/apache/pulsar/blob/6382be84ae023edeae1ccc996d70391e8b01aec4/pulsar-broker/src/test/java/org/apache/pulsar/broker/auth/MockedPulsarServiceBaseTest.java#L87

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApiTest2.java [hunk 2]
```diff
@@ -2231,4 +2232,170 @@ public void testFailedUpdatePartitionedTopic() throws Exception {
...
+        PartitionedTopicStats topicStats = admin.topics().getPartitionedStats(topic, true);
+        assertEquals(topicStats.getPartitions().size(), 2);
+        assertEquals(topicStats.getPartitions().values().stream().mapToInt(e -> e.getPublishers().size()).sum(), 2);
+        assertEquals(topicStats.getPartitions().values().stream().map(e -> e.getPublishers().get(0).getProducerName()).distinct().count(), 2);
+        assertEquals(topicStats.getPublishers().size(), 2);
+        topicStats.getPublishers().forEach(p -> assertTrue(p.isPartialProducerSupported()));
+    }
+
+    @Test
+    public void testPartitionedStatsAggregationByProducerNameNonPersistent() throws Exception {
```
Comment (massakam) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApiTest2.java:2320]: The only difference between this and `testPartitionedStatsAggregationByProducerName()` seems to be whether the topic is persistent or non-persistent. Wouldn't we be able to cover both test cases with one method using `@DataProvider` etc?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/GracefulExecutorServicesShutdownTest.java [hunk 3]
```diff
@@ -152,7 +153,8 @@ public void shouldTerminateWhenFutureIsCancelled() throws InterruptedException,
         // when
         shutdown.shutdown(executorService);
         CompletableFuture<Void> future = shutdown.handle();
-        awaitingTerminationEntered.await();
+        // waiting to start awaitTermination
+        verify(executorService, timeout(100).atLeastOnce()).awaitTermination(anyLong(), any());
```
Comment (massakam) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/GracefulExecutorServicesShutdownTest.java:157]: I don't know why this change was made, but isn't `awaitingTerminationEntered` no longer needed?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/PartitionedProducerImpl.java [hunk 4]
```diff
@@ -145,42 +147,67 @@ private void start(List<Integer> indexList) {
         AtomicReference<Throwable> createFail = new AtomicReference<Throwable>();
         AtomicInteger completed = new AtomicInteger();
 
-        for (int partitionIndex : indexList) {
-            createProducer(partitionIndex).producerCreatedFuture().handle((prod, createException) -> {
-                if (createException != null) {
-                    setState(State.Failed);
-                    createFail.compareAndSet(null, createException);
+        final BiConsumer<Boolean, Throwable> after = (failFast, createException) -> {
```
Comment (massakam) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/PartitionedProducerImpl.java:150]:
```suggestion
        final BiConsumer<Boolean, Throwable> afterCreatingProducer = (failFast, createException) -> {
```

**Comment by massakam** [comment]
Do we need to update the documentation? For example:
https://github.com/apache/pulsar/blob/f60a2a57ae84c7f18cf6f9ea60ae2b52c3e29aec/site2/docs/admin-api-topics.md#get-stats

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/GracefulExecutorServicesShutdownTest.java (hunk 3 above)
Comment (equanz) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/GracefulExecutorServicesShutdownTest.java:157]: I didn't notice that the problem was already fixed by `awaitingTerminationEntered`. I'll remove this changes.

**Comment by equanz** [comment]
I think I'll create a new PR and update docs about both https://github.com/apache/pulsar/pull/10279 and https://github.com/apache/pulsar/pull/12401 at new PR. Is it okay?

> Do we need to update the documentation? For example:
> https://github.com/apache/pulsar/blob/f60a2a57ae84c7f18cf6f9ea60ae2b52c3e29aec/site2/docs/admin-api-topics.md#get-stats
> 
> https://github.com/apache/pulsar/pull/12401#issuecomment-962912476

**Comment by equanz** [comment]
@massakam Addressed your comments. PTAL

**Review by massakam — APPROVED** [review]
LGTM

File: pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/stats/TopicStatsImpl.java [hunk 5]
```diff
@@ -90,7 +95,12 @@
 
     /** List of connected publishers on this topic w/ their stats. */
     @Getter(AccessLevel.NONE)
-    public List<PublisherStatsImpl> publishers;
+    @Setter(AccessLevel.NONE)
+    private List<PublisherStatsImpl> publishers;
+
+    @Getter(AccessLevel.NONE)
+    @Setter(AccessLevel.NONE)
+    private Map<String, PublisherStatsImpl> publishersMap;
```
Comment (eolivelli) [inline pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/stats/TopicStatsImpl.java:103]: Is this struct

**Review by eolivelli — CHANGES_REQUESTED** [review]
Overall the change makes sense to me.
But I can't remember about the discussion about the wire protocol change.

Is there another way to achieve this without changing

File: pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/stats/TopicStatsImpl.java (hunk 5 above)
Comment (equanz) [reply pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/stats/TopicStatsImpl.java:103]: In my understanding, no. This field is just for internal use. If the jackson-databind selialize this POJO to JSON, the field will be returned as publishers through `TopicStatsImpl#getPublishers`. Also, deserialize it, the field will be separated to publishers and publishersMap through `TopicStatsImpl#setPublishers`.

This because PublisherStatsImpl#partialProducerSupported is serialized/deserialized by jackson-databind.

```
% ./bin/pulsar-admin topics partitioned-stats persistent://public/default/pt --per-partition
{
  "msgRateIn" : 0.0,
  "msgThroughputIn" : 0.0,
  "msgRateOut" : 0.0,
  "msgThroughputOut" : 0.0,
  "bytesInCounter" : 488,
  "msgInCounter" : 8,
  "bytesOutCounter" : 0,
  "msgOutCounter" : 0,
  "averageMsgSize" : 0.0,
  "msgChunkPublished" : false,
  "storageSize" : 488,
  "backlogSize" : 0,
  "offloadedStorageSize" : 0,
  "lastOffloadLedgerId" : 0,
  "lastOffloadSuccessTimeStamp" : 0,
  "lastOffloadFailureTimeStamp" : 0,
  "publishers" : [ {
    "msgRateIn" : 0.0,
    "msgThroughputIn" : 0.0,
    "averageMsgSize" : 0.0,
    "chunkedMessageRate" : 0.0,
    "producerId" : 0,
    "partialProducerSupported" : true,
    "producerName" : "standalone-0-0"
  } ],
  "waitingPublishers" : 0,
  "subscriptions" : { },
  "replication" : { },
  "nonContiguousDeletedMessagesRanges" : 0,
  "nonContiguousDeletedMessagesRangesSerializedSize" : 0,
  "compaction" : {
    "lastCompactionRemovedEventCount" : 0,
    "lastCompactionSucceedTimestamp" : 0,
    "lastCompactionFailedTimestamp" : 0,
    "lastCompactionDurationTimeInMills" : 0
  },
  "metadata" : {
    "partitions" : 2
  },
  "partitions" : {
    "persistent://public/default/pt-partition-1" : {
      "msgRateIn" : 0.0,
      "msgThroughputIn" : 0.0,
      "msgRateOut" : 0.0,
      "msgThroughputOut" : 0.0,
      "bytesInCounter" : 0,
      "msgInCounter" : 0,
      "bytesOutCounter" : 0,
      "msgOutCounter" : 0,
      "averageMsgSize" : 0.0,
      "msgChunkPublished" : false,
      "storageSize" : 0,
      "backlogSize" : 0,
      "offloadedStorageSize" : 0,
      "lastOffloadLedgerId" : 0,
      "lastOffloadSuccessTimeStamp" : 0,
      "lastOffloadFailureTimeStamp" : 0,
      "publishers" : [ ],
      "waitingPublishers" : 0,
      "subscriptions" : { },
      "replication" : { },
      "deduplicationStatus" : "Disabled",
      "nonContiguousDeletedMessagesRanges" : 0,
      "nonContiguousDeletedMessagesRangesSerializedSize" : 0,
      "compaction" : {
        "lastCompactionRemovedEventCount" : 0,
        "lastCompactionSucceedTimestamp" : 0,
        "lastCompactionFailedTimestamp" : 0,
        "lastCompactionDurationTimeInMills" : 0
      }
    },
    "persistent://public/default/pt-partition-0" : {
      "msgRateIn" : 0.0,
      "msgThroughputIn" : 0.0,
      "msgRateOut" : 0.0,
      "msgThroughputOut" : 0.0,
      "bytesInCounter" : 488,
      "msgInCounter" : 8,
      "bytesOutCounter" : 0,
      "msgOutCounter" : 0,
      "averageMsgSize" : 0.0,
      "msgChunkPublished" : false,
      "storageSize" : 488,
      "backlogSize" : 0,
      "offloadedStorageSize" : 0,
      "lastOffloadLedgerId" : 0,
      "lastOffloadSuccessTimeStamp" : 0,
      "lastOffloadFailureTimeStamp" : 0,
      "publishers" : [ {
        "accessMode" : "Shared",
        "msgRateIn" : 0.0,
        "msgThroughputIn" : 0.0,
        "averageMsgSize" : 0.0,
        "chunkedMessageRate" : 0.0,
        "producerId" : 0,
        "partialProducerSupported" : true,
        "metadata" : { },
        "address" : "/127.0.0.1:64336",
        "producerName" : "standalone-0-0",
        "connectedSince" : "2021-11-15T19:21:23.335+09:00",
        "clientVersion" : "2.9.0-SNAPSHOT"
      } ],
      "waitingPublishers" : 0,
      "subscriptions" : { },
      "replication" : { },
      "deduplicationStatus" : "Disabled",
      "nonContiguousDeletedMessagesRanges" : 0,
      "nonContiguousDeletedMessagesRangesSerializedSize" : 0,
      "compaction" : {
        "lastCompactionRemovedEventCount" : 0,
        "lastCompactionSucceedTimestamp" : 0,
        "lastCompactionFailedTimestamp" : 0,
        "lastCompactionDurationTimeInMills" : 0
      }
    }
  }
}
```

File: pulsar-common/src/main/proto/PulsarApi.proto [hunk 6]
```diff
@@ -485,6 +485,8 @@ message CommandProducer {
     optional uint64 topic_epoch = 11;
 
     optional bool txn_enabled = 12 [default = false];
+
+    optional bool partial_producer_supported = 13 [default = false];
```
[the remark this answers is not in the document]
Comment (equanz) [reply pulsar-common/src/main/proto/PulsarApi.proto:489]: Hmm... First, we didn't discuss about the wire protocol change. So, I'll also send a new mail to [the thread](https://mail-archives.apache.org/mod_mbox/pulsar-dev/202102.mbox/%3c5490ED12-C172-4AD6-8E77-2E70C66C9793@yahoo-corp.jp%3e).

Currently, partitioned producer can't aggregate by any key such as cnx, producerId, producerName, and so on. I think we need to add any aggregation system. Therefore, added new aggregation policy as producerName (with client side implementation).

New protocol field [partial_producer_supported](https://github.com/apache/pulsar/pull/12401/files#diff-f29399fed32e0916cf28452ba71078a3ae5ed77bbaef9f52a925165d8ee66cfdR489) is not used for stats aggregation. It is used for backward compatibility.

In my understanding, if introduce new stats aggregation key to client side, need a way to determine whether the feature is enabled at client side. For example, whether the producer has specific field or metadata, the version (e.g. protocol version) is greater than threshold, etc.

Of course, if we can introduce aggregation feature without adding any new key or implementations from client side, we can support the feature not only new client but also old one.

**Comment by equanz** [comment]
org.apache.pulsar.broker.service.ReplicatorTest

**Comment by equanz** [comment]
@eolivelli @k2la @massakam 
I add the new config key `aggregatePublisherStatsByProducerName` to follow [the comment](https://lists.apache.org/thread/g5b9wh9kb0k2gq2gb4rxbf6yz0gg18m3).
If true and the client supports partial producer, then aggregate publisher stats of PartitionedTopicStats by producerName (proposed behavior). Otherwise, aggregate it by list index (existing behavior).

PTAL

File: pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/stats/NonPersistentTopicStatsImpl.java [hunk 7]
```diff
@@ -57,7 +62,11 @@
 
     @JsonProperty("publishers")
     public List<NonPersistentPublisherStats> getNonPersistentPublishers() {
-        return (List<NonPersistentPublisherStats>) nonPersistentPublishers;
+        return Stream.concat(nonPersistentPublishers.stream()
+                                .sorted(Comparator.comparing(NonPersistentPublisherStats::getProducerName)),
```
Comment (eolivelli) [inline pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/stats/NonPersistentTopicStatsImpl.java:66]: sorting here may be a problem for users.
because users will tend to think that this list is always sorted by producerName, but there is documentation about this, and stuff like this tend to break in future refactors.

do you really need this list to be sorted ?

**Review by eolivelli — COMMENTED** [review]
I left one last comment

PTAL

File: pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/stats/NonPersistentTopicStatsImpl.java (hunk 7 above)
Comment (equanz) [reply pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/stats/NonPersistentTopicStatsImpl.java:66]: > do you really need this list to be sorted ?

I don't strongly need the sorted order. It is for a consistent reason. Currently, publishers have some constant order even if they get stats in many times.
I'll remove the reordering operation if we don't have to support consistent orders.

Comment (equanz) [reply pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/stats/NonPersistentTopicStatsImpl.java:66]: @eolivelli

**Review by eolivelli — APPROVED** [review]
+1

Great work
